kv/batch.go

```diff
@@ -22,99 +22,93 @@ import (
 	"github.com/cockroachdb/cockroach/keys"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/util"
+	"github.com/gogo/protobuf/proto"
 	"golang.org/x/net/context"
 )
 
-// truncate restricts all contained requests to the given key range.
-// Even on error, the returned closure must be executed; it undoes any
-// truncations performed.
-// All requests contained in the batch are "truncated" to the given
+var emptySpan = roachpb.Span{}
+
+// truncate restricts all contained requests to the given key range
+// and returns a new BatchRequest.
+// All requests contained in that batch are "truncated" to the given
 // span, inserting NoopRequest appropriately to replace requests which
 // are left without a key range to operate on. The number of non-noop
-// requests after truncation is returned along with a closure which
-// must be executed to undo the truncation, even in case of an error.
-// TODO(tschottdorf): Consider returning a new BatchRequest, which has more
-// overhead in the common case of a batch which never needs truncation but is
-// less magical.
-func truncate(br *roachpb.BatchRequest, rs roachpb.RSpan) (func(), int, error) {
-	truncateOne := func(args roachpb.Request) (bool, []func(), error) {
+// requests after truncation is returned.
+func truncate(ba roachpb.BatchRequest, rs roachpb.RSpan) (roachpb.BatchRequest, int, error) {
+	truncateOne := func(args roachpb.Request) (bool /* omit? */, roachpb.Span, error) {
```

Comment:
Yes, that's what I mean. When there's an error, `ok` should clearly be false, both because it's the zero value and semantically correct. If it's `omit`, `true` is arguably more correct but it's unusual to return a non-zero value with an error.